### ISSUE TITLE
gulp watch: stream.js throws error

### ISSUE THREAD

**Issue by erikwestlund** [laravel/elixir#626]
Gulp runs fine, but when using gulp watch I get the following error:

```
stream.js:74
      throw er; // Unhandled stream error in pipe.
      ^

Error: ENOENT: no such file or directory, stat '.../public/build/assets/js/admin-a30ce533d9.js'
    at Error (native)
```

I found this: https://github.com/ben-eb/gulp-cssnano/issues/23

But am not sure how to proceed.

**Comment by brnmonteiro** [comment]
any solution? here is the same!

**Comment by MichMich** [comment]
Same here!

**Comment by laracasts** [comment]
Show me your Gulpfile so I have some way to reproduce this.

**Comment by MichMich** [comment]
```javascript
const elixir = require("laravel-elixir");
require("laravel-elixir-vue");

elixir(function(mix) {
    //Fontend
    mix.sass(["frontend.scss"], "public/css/frontend.css");
    mix.webpack("frontend.js", "public/js/frontend.js");

    //Backend
    mix.sass([
        "../../../node_modules/medium-editor/dist/css/medium-editor.css",
        "../../../node_modules/medium-editor/dist/css/themes/default.css",
        "backend.scss"
    ], "public/css/backend.css");
    mix.webpack("backend.js", "public/js/backend.js");

    //Copy files
    mix.copy("node_modules/font-awesome/fonts", "public/build/fonts");
    mix.copy("resources/assets/sass/vendor/inspinia/patterns", "public/build/css/patterns");

    //Make versions
    mix.version(["css/frontend.css","js/frontend.js","css/backend.css","js/backend.js"]);
});
```

**Comment by nacr-dev** [comment]
Same problem here.

**Comment by danielsdeboer** [comment]
I've had this problem too, I eventually "fixed" it by chunking my gulpfile in different bits with yargs, which is obviously not ideal.

Usually what happens for me when longer-running processes (Sass & Webpack for instance) seem to be running asynchronously on the initial compile and trigger the version task again and again.

**Comment by erikwestlund** [comment]
Here is mine: 

```
var gulp = require('gulp');
var elixir = require('laravel-elixir');
require('laravel-elixir-vue');


Elixir.ready(() => {

    Elixir.webpack.mergeConfig({
        babel: {
            presets: ['es2015'],
            plugins: ['transform-runtime'],
        },
        module: {
            loaders: [{
                test: /\.html$/,
                loader: 'vue-html'
            }]
        }
    });

});

/**
 * Copy any needed files.
 */
gulp.task('copyfiles', function() {

    // admin panel + plugins
    gulp.src('resources/assets/images/admin/**')
        .pipe(gulp.dest('public/build/assets/css/'));

    //font awesome
    gulp.src('vendor/bower_dl/font-awesome/fonts/**')
        .pipe(gulp.dest('public/build/assets/fonts'));

});

/*
 * Mix the assets
 */

elixir(function(mix) {

    // main
    mix.sass('resources/assets/sass/main.scss', 'public/assets/css/main.css');
    mix.webpack('resources/assets/js/main.js', 'public/assets/js/main.js');

    // admin 
    mix.sass('resources/assets/sass/admin.scss', 'public/assets/css/admin.css')
    mix.webpack('admin.js', 'public/assets/js/admin.js');

    // versions for cache-busting
    mix.version([
        'assets/js/main.js',
        'assets/css/main.css',
        'assets/js/admin.js',
        'assets/css/admin.css'
    ]);

});
```

**Comment by StevenBock** [comment]
It probably has to do with the multiple webpack entrypoints as I've gotten the same message in #598.

**Comment by lukepolo** [comment]
I also get this on occasion, re-running gulp watch seems to fix it though.

**Comment by gregrobson** [comment]
I'm getting something similar which might be related. Very occasionally the script will work. I've tried putting everything inside the same `elixir(function(mix) {` block.

I've run console as admin and have given every permission under the sun to the folder - still no joy.

I can only think that something async hasn't finished writing the file before the versioning is finished???

N.B. I'm no expert on JavaScript/Gulp! I many have overlooked something obvious!

```
┌───────────────┬──────────────────────────┬────────────────────────┬──────────────┐
│ Task          │ Summary                  │ Source Files           │ Destination  │
├───────────────┼──────────────────────────┼────────────────────────┼──────────────┤
│ mix.version() │ 1. Copying Source Maps   │ public\compile\all.js  │ public\build │
│               │ 2. Versioning            │ public\compile\all.css │              │
│               │ 3. Rewriting File Paths  │                        │              │
│               │ 4. Saving to Destination │                        │              │
└───────────────┴──────────────────────────┴────────────────────────┴──────────────┘
[11:07:04] Finished 'version' after 31 ms
stream.js:74
      throw er; // Unhandled stream error in pipe.
      ^

Error: EPERM: operation not permitted, open 'C:\nginx\web\psapp.com.dev\public\build\compile\all-cc6fdd75e3.js'
    at Error (native)
```

My gulp file is this:
```js
const elixir = require('laravel-elixir');

require('laravel-elixir-vue');

/*
 |--------------------------------------------------------------------------
 | Elixir Asset Management
 |--------------------------------------------------------------------------
 |
 | Elixir provides a clean, fluent API for defining some basic Gulp tasks
 | for your Laravel application. By default, we are compiling the Sass
 | file for our application, as well as publishing vendor resources.
 |
 */

// elixir(mix => {
//     mix.sass('app.scss')
//        .webpack('app.js');
// });

// Sources relative to: resources/assets/css
// Results go into public/css/all.css:
elixir(function(mix) {
    mix.styles([
        'imports.css', // @import statements must come first in execution
        'bootstrap/bootstrap_3.3.7.css',
        'font-awesome_4.6.3.css',
        'sweetalert2/sweetalert2_4.2.6.css',
        'bootstrap-datepicker/bootstrap-datepicker3_1.6.4.css',
        'inspinia/animate.css',
        'inspinia/style.css',
        'custom.css', // Should be last!
    ], 'public/compile');
});

elixir(function(mix) {
    mix.scripts([
        // Vue
        'vue/vue_2.0.0RC5.js',
        'vue/vue-resource_0.9.3.js',
        // 'vue/vue.common.js',
        'jquery-3.1.0.js',
        'promise-7.0.4.js',
        'bootstrap_3.3.7.js', // relies on jQuery
        'inspinia/inspinia.js',
        'inspinia/plugins/metisMenu/jquery.metisMenu.js', // Auto-collapsing of side menu
        'inspinia/plugins/slimscroll/jquery.slimscroll.js', // Mini-scrollbars in DIVs
        'moment/moment_2.14.1.js',
        'sweetalert2/sweetalert2_4.2.6.js',
        'bootstrap-datepicker/bootstrap-datepicker_1.6.4.js',
        // 'jquery-validate/jquery.validate-1.15.1.js',
        // 'jquery-validate/additional-methods-1.15.1.js',
        'custom.js',
    ], 'public/compile');
});

// Cache-busts the CSS/JS files so that it gets a new hash whenever
// the source details change.
elixir(function(mix) {
    mix.version([
        'public/compile/all.js',
        'public/compile/all.css'
    ]);
});

elixir(function(mix) {
    mix.browserSync({
        proxy: 'psapp.com.dev'
    });
});
```

**Comment by nacr-dev** [comment]
@JeffreyWay just to say that I'm using browserify and I have the same problem.

**Comment by MichMich** [comment]
Still no solution here. `gulp` works fine. `gulp watch` still doesn't.

**Comment by nacr-dev** [comment]
any one has any idea about how to resolve this?

**Comment by erikwestlund** [comment]
Nuno,

For me, the issue is having more than one call in elixir. For example, I
have mix calls for CSS and JS, both front end and main.

Usually, I'm only working on one of those.  I just comment out the others
and run gulp watch and it works.

Not a lng-term solution but it saves you some alt-tab + gulp + wait 5
seconds in the terminal

**Comment by nacr-dev** [comment]
I Erik,

That's a good Point tks.

Hope your having a nice day :)

**Comment by jakedowns** [comment]
just wanted to chime in that I'm having this issue as well after recently upgrading to the latest version of elixir. (Here's my [package.json](https://gist.github.com/jakedowns/a902732e69c6bdf43df836154ae200a3))

node v4.6.0
npm 2.15.9

my gulpfile calls:
mix.sass            4x
mix.browserify  2x 
mix.scripts        2x 
mix.version       1x

FWIW deleting the destination directory seems to temporarily resolve the issue?

I'm also working within a Vagrant environment on a Windows host machine, if that makes any difference. Same thing happens if I run gulp from within Vagrant or from Host.

**Comment by erikwestlund** [comment]
One tip:

When I run:

gulp && gulp watch


I don't have the issues even with multiple mix commands.

However, if I cancel out and do something else in the terminal and just run
"gulp watch"

i then often run into issues.


At least I think that's what's happening :)

**Comment by jakedowns** [comment]
Interesting... I usually start with a `gulp; gulp watch` out of habit. I'll try with `&&` instead

**Comment by erikwestlund** [comment]
Not sure that would make a difference as I think they are functionally
equivalent.

I wonder if the versioning may be causing the problems?

**Comment by jakedowns** [comment]
Yeah, still doing the same for me. Need to move on for now. Guess I'll just manually run `gulp` and hope watcher works again soon. :)

**Comment by Dronnikkl** [comment]
Same here

**Comment by lbausch** [comment]
In my case it was sufficient to delete the `public/build` directory before running `gulp watch` so I modified my `package.json` accordingly:

```json
 "scripts": {
    "prod": "gulp --production",
    "dev": "rm -rf public/build && gulp watch"
  },
```

**Comment by jakedowns** [comment]
@lbausch hrm... gave this a shot and it didn't seem to resolve it for me. deleting the output dirs does fix it so a single `gulp` will work as expected. but `gulp watch` still throws I/O error. Even if I add a custom gulp.watch task that includes a rm using gulp-shell.

Tried updating to latest node (v6.7.0) & the issue persists. Starting to think this might be related to my use of the vagrant winnfsd plugin?

**Comment by MichMich** [comment]
@jakedowns Nope, same issue on a regular Mac install.

**Comment by jakedowns** [comment]
@MichMich thanks for the confirmation, that's good to know. Maybe related to https://github.com/JeffreyWay/laravel-elixir-webpack-official/issues/3 ?

**Comment by jakedowns** [comment]
_update_ thought https://github.com/JeffreyWay/laravel-elixir-webpack-official/commit/ac6aa6507252a74eb9c970548680c4117b6de27d might fix this up, but no luck. still getting stream.js:74 error, even when I comment out my mix.webpack calls. (tried browserify, webpack & rollup. happens across the board)

![](https://d3vv6lp55qjaqc.cloudfront.net/items/2c281n1P38000i3v1G1r/Image%202016-10-06%20at%2011.53.12%20AM.png?X-CloudApp-Visitor-Id=88002&v=5216f066)

**Comment by unitedworx** [comment]
I am having the same issue as well and seems that having multiple mix.weppack() and mix.sass() is cusign this!

**Comment by JeffreyWay** [comment]
Multiple mix.sass() is fine, but you all should use only one mix.webpack() until we push a commit to fix a couple things for that.

**Comment by unitedworx** [comment]
Thanks for the reply @JeffreyWay at the moment we need to painfully work running gulp manually every time.

Multiple webpack calls are definitely needed when you have a scenario where you have complexity different setup for the actual website and its admin area. 

hopefully a fix will come soon

**Comment by cuonghuynh** [comment]
I got same error when I run gulp on laravel 5.3 (newly project). 

![](https://i.gyazo.com/deb03221d24a79dc41aba6bc201d75fc.png)

**Comment by jakedowns** [comment]
@JeffreyWay FYI, same thing happens when I comment out my calls to mix.webpack entirely.

**Comment by bbashy** [comment]
This is only an issue with a newer version of npm itself. I have multiple .webpack() in gulpfile.js but doing `npm i -g npm@latest` broke it and that's why I'm here now :P

**Comment by briano-jabico** [comment]
I am also having this problem. I recently upgraded to Node 6.9.0 from 0.12.7 and started using `mix.webpack()`. When I call `mix.scripts()` instead of `mix.webpack()`, I do not have this problem.

It happens with `gulp watch`, not with `gulp`. It does not happen every time; 30 minutes ago it was happening maybe 80% of the time, but right now I'm having no problems. It can happen with both the initial `gulp watch` and when I modify a JS file once it is successfully watching, forcing a webpack recompile.

I only have one `elixir(function() { ... });` call. I have many mix calls within that:
``` JavaScript
elixir(function(mix) {

    mix.less(/* ... */);
    mix.sass(/* ... */);
    mix.scripts(/* ... */);
    mix.scripts(/* ... */);
    mix.styles(/* ... */);
    mix.sass(/* ... */);

    mix.webpack(/* ... */);    // <-- This causes the error sometimes
    // mix.scripts(/* ... */); // <-- This never causes the error

    mix.scripts(/* ... */);
    mix.styles(/* ... */);
    // inFiles is declared elsewhere
    Object.keys(inFiles.copyFiles).forEach(function(key) {
        mix.copy(key, inFiles.copyFiles[key]);
    })
    mix.version(/* ... */); // About 8 files, plus public/fonts/**/*, plus public/images/**/*

    mix.browserSync({
        proxy: /* ... */,
        files: [
            'app/**/*.php',
            'resources/**/*.php',
            'public/build/rev-manifest.json',
        ],
    });

});
```

**Comment by bhulsman** [comment]
This was an issue in version 5.x of Elixir and it indeed still is. It looks like it starts with versioning before compiling of scripts has finished.

I would very much like to use Gulp with Elixir on all our projects, but at the moment just running LESS and Uglify File Watchers in PhpStorm is a lot faster. Especially when `gulp watch` stops working once in a while and you need to restart it.

**Comment by pmall** [comment]
I gave up trying to reproduce this error. There is no logic behind this, except it get fixed when you run gulp, then gulp watch. Hope it will get fixed someday.

**Comment by AshboDev** [comment]
@pmall That doesn't fix it for me. I've tried `gulp` and `gulp watch` as well as `gulp && gulp watch` both of which throw the error. I'm having to manually gulp *every* single time. I've been working on this project for a few days now, and it's a pain in the rear. I'm gonna try updating npm again as it only broke when I updated it.

**Comment by Frondor** [comment]
I'm also facing this problem. Running `gulp && gulp watch` *sometimes* just works

**Comment by pmall** [comment]
Yes it does not work anymore for me, for no reason. Elixir is simply unusable for development right now.

**Comment by bbashy** [comment]
Definitely something up with it. Multiple people on different projects on different machines.

**Comment by bobbybouwmann** [comment]
@JeffreyWay How far are you with this?
> Multiple mix.sass() is fine, but you all should use only one mix.webpack() until we push a commit to fix a couple things for that.

**Comment by baadillahnabil** [comment]
anything new on this? i really hope this will fixed soon :)

**Comment by briano-jabico** [comment]
@bobbybouwmann I commented earlier but I'd like to make sure this stands out: I was having the problem with just one `mix.webpack(/* ... */);` call, though I was passing an array of files to that one call.

**Comment by arvindeleon** [comment]
Bump on this one.

**Comment by bbashy** [comment]
Is there a problem with this being fixed? Is there a different way to go about it?

**Comment by lukepolo** [comment]
I believe its this : #681 , but not sure how they are going to fix it

**Comment by tobros91** [comment]
Any fix to this? gulp watch crashes 9/10 times for me. 


    
![webpack](https://cloud.githubusercontent.com/assets/4538330/20106464/93081c6c-a5d5-11e6-959c-f868f156dd02.png)

    mix.sass(['admin/app.scss'], 'public/css/admin.css')
   		.sass(['clinic/app.scss'], 'public/css/clinic.css')
   		.sass(['card/app.scss'], 'public/css/card.css')

    mix.webpack('admin/app.js', './public/js/admin.js')
        .webpack('clinic/app.js', './public/js/clinic.js')
        .webpack('card/app.js', './public/js/card.js');

    mix.copy([
    	'node_modules/bootstrap-sass/assets/fonts/bootstrap/',
    	'node_modules/font-awesome/fonts/'
    	],
    	'public/fonts/bootstrap'); 
  
	mix.version(['css/admin.css', 'css/clinic.css', 'css/card.css', 'js/admin.js', 'js/clinic.js', 'js/card.js']);

**Comment by korneliuskristianr** [comment]
I've same problem with elixir versioning

**Comment by unitedworx** [comment]
I no longer use gulp watch since it simply does not work. I run gulp all the time manually but it's a pain if you work with a vue/js. Hopefully a fix is inbound soon

**Comment by sunscreem** [comment]
I'm using sublime text 3. I excluded /public/build from my project and this has solved it for me.

I suspect sublime was locking some files for reindexing and screwing things up.

**Comment by daverdalas** [comment]
@sunscreem I checked this solution with PhpStorm and the problem persists :(

**Comment by msamec** [comment]
Like @lbausch already mentioned it helps if you delete build folder before you execute `gulp watch`. At least it works for me. I know it is not a best solution but I'd rather delete build once a day and have `gulp watch` working than have to run `gulp` every time.

**Comment by brnmonteiro** [comment]
I delete `rev-manifest.json` and works.

**Comment by arvindeleon** [comment]
@brnmonteiro thank you it works for me also, deleted rev-manifest.json and run gulp watch while on ssh (homestead)

**Comment by sailingdeveloper** [comment]
I now use `rm -rf public/build/rev-manifest.json && gulp watch` instead of `gulp watch`. Doesn't seem ideal.

**Comment by unitedworx** [comment]
I do rm -Rf public/build && gulp watch as well :) 
its a workaround that solved my problem until this is properly fixed.

**Comment by mistre83** [comment]
Since yesterday i havent any problems with gulp watch, and today i have this error.

What is changed: by yesterday i have installed ```async``` and ```chart.js```.

But, if i do this sequence, ```gulp watch``` works!

1 - Remove ```node_modules``` and chart.js from ```package.json```
2 - Re install with ```npm install```
3 - Run ```gulp watch``` -> NO ERROR
4 - Add again chart.js and re run ```npm install```
5 - Run ```gulp watch``` -> NO ERROR

The only difference is that i install ```chart.js``` in a step later, but the library are the same!

**Comment by maki10** [comment]
Try to run `gulp` outside homestead. That help to me.

**Comment by mistre83** [comment]
i even run gulp outside..

**Comment by JeramyNS** [comment]
Same issue over here when including versioning; always works without it.

After some fiddling I decided to manually include the build folder which seems to of resolved the issue for me. Not sure why it makes a difference but didn't really look into it.
Does it resolve the issue for anyone else?

My Example: 
```
  mix.version([
        'js/main.js',
        'css/main.css',
        'css/public.css',
        'css/built.css',
        'js/built.js'
    ], './public/build'); // Manually include build folder
```

**Comment by RonMelkhior** [comment]
@JeramyNS That has just resolved it for me as well. Thanks!

**Comment by RonMelkhior** [comment]
Just fixed my issue by updating to Node v6.9.4 on Windows.

**Comment by encodia** [comment]
Fixed using @JeramyNS approach. It seems that it works.

**Comment by eleftrik** [comment]
@RonMelkhior Just updated to Node v6.9.4 on Windows, but problem hasn't been solved.

**Comment by RonMelkhior** [comment]
@eleftrik That's weird. Hope this gets fixed.

Otherwise, you can look forward to Laravel Mix which is a pretty quick upgrade (took me ~15 minutes).

**Comment by JosephBywater** [comment]
I had this issue because I had a clean task. This was being run asynchronously with all of my other tasks causing the error. To fix, I removed the clean task as it wasn't really needed. However, you could just add the task as a dependency so that the tasks happen in order.

**Comment by RonMelkhior** [comment]
@JosephBywater 

You're right, I tested it on another machine that was experiencing the issue.

Switched from
```js
elixir(mix => {
    mix
    .sass('vendor.scss', 'public/assets/css/vendor.css')
    .sass('app.scss', 'public/assets/css/app.css')
    .webpack('vendor.js', 'public/assets/js/vendor.js')
    .webpack('app.js', 'public/assets/js/app.js')
    .copy('resources/assets/img', 'public/assets/img')
    
    mix.version(['assets/css/*.css', 'assets/js/*.js', 'assets/img/*.png']);
});
```
to
```js
elixir(mix => {
    mix
    .sass('vendor.scss', 'public/assets/css/vendor.css')
    .sass('app.scss', 'public/assets/css/app.css')
    .webpack('vendor.js', 'public/assets/js/vendor.js')
    .webpack('app.js', 'public/assets/js/app.js')
    .copy('resources/assets/img', 'public/assets/img')
    .version(['assets/css/*.css', 'assets/js/*.js', 'assets/img/*.png']);
});
```

And it now works beautifully. Thank you!

**Comment by MichMich** [comment]
Anyone still running into issues: you might want to switch to https://github.com/JeffreyWay/laravel-mix
This will be the default for L5.4 and is WebPack based (goodbye Gulp).

I haven't had any issues since I've switched to Mix.

**Comment by felipedinni** [comment]
I make script in c++, and ecute in source folder... normally works until the third interaction
compile: g++ name.cpp -o gulp
`
#include <iostream>
#include <cstdlib>

using namespace std;

int main() {
	int cont = 0;
	while (true) {
		cout << ++cont << endl << endl << endl;
		system("gulp && gulp watch");
		system("cls");
	}

	return 0;
}
`

**Comment by mabasic** [comment]
Started happening to me lately.. Can't explain it. Happens at random times. Can't tell what solution works because sometimes they work, sometimes they don't.

**Comment by marceloch2** [comment]
Well, after implementing the hot-reload the problem didn't happen anymore.

mix.browserSync({
        proxy: 'ireland-planner.lara',
        files: [
            elixir.config.appPath + '/**/*.php',
            elixir.config.get('public.css.outputFolder') + '/**/*.css',
            elixir.config.get('public.versioning.buildFolder') + '/rev-manifest.json',
            'resources/views/**/*.php'
        ],
    });

**Comment by TechyTimo** [comment]
I also have the same issue. My hack has been to clear the lines in build/rev-manifest.json every time before I do gulp watch. Leaving it with as an empty object {}
